db/fdb_fend.c

```diff
@@ -1960,7 +1960,13 @@ int fdb_cursor_move_master(BtCursor *pCur, int *pRes, int how)
     }
 
 search:
-    __lock_wrlock_shared(fdb);
+    //__lock_wrlock_shared(fdb);
+
+    /* At this point fdb->users already is increased to 2 by this thread and
+     * there can be multiple threads waiting for users to decrease.
+     * One of them gets write lock while others wait for the write lock,
+     * so this thread will never get a read lock */
+    Pthread_rwlock_wrlock(&fdb->h_rwlock);
```

Comment:
This is not right, it will serialize access to sqlite master.   users is incremented to allow lock free access, but not free the structure!  This is not a source of deadlock

Reply:
Close to 1000 other thread  in _add_table_and_stats_fdb one of which has wrlock and other waiting at https://github.com/bloomberg/comdb2/blob/master/db/fdb_fend.c#L1476 , the one which has wrlock fails and gets back to the wait, leaving the lock for one of the 1000 theads. This thead never gets opportunity to proceed.

Reply:
How do we allow 1000 sqlite engines?

Reply:
Only one sqlite engine will get to populate a fdb cache, the rest are simply waiting for the result.  This is by design.  No need to have every engine try to grab same remote schema information.

Reply:
Its lua trigger threads here, a new thread is created if the previous thread is stuck and not consuming queue
https://github.com/bloomberg/comdb2/blob/master/db/trigger.c#L132

Reply:
We found the deadlock culprit and opened a new PR for it https://github.com/bloomberg/comdb2/pull/2455.
What happens it, due to the fact that sqlite_master has no entry in fdb cache, the sqlite table hash remains unpopulated during the initial fdb creation.  When sqlite later on tries to access sqlite_stats, as part of the fdb creation and table attachment, it has no entry for sqlite_stats and trigger a nested fdb creation inside the fdb creation process.  This obviously deadlocks since the access is exclusive.
Lets remove the changes for locking in remsql and keep only the trigger limit for this PR.